File: coding-patterns/16-binary-search/119.py
# Search in a Sorted Infinite Array (Medium)

import math


class ArrayReader:

  def __init__(self, arr):
    self.arr = arr

  def get(self, index):
    if index >= len(self.arr):
      return math.inf
    return self.arr[index]


class Solution:
  def searchInfiniteSortedArray(self, reader, key):
    start, end = 0, 1
    while reader.get(end) < key:
      newStart = end + 1
      end += (end - start + 1) * 2
      # increase to double the bounds size
      start = newStart

    return self.binary_search(reader, key, start, end)


  def binary_search(self, reader, key, start, end):
    while start <= end:
      mid = start + (end - start) // 2
      if key < reader.get(mid):
        end = mid - 1
      elif key > reader.get(mid):
        start = mid + 1
      else:  # found the key
        return mid

    return -1

File: coding-patterns/16-binary-search/119.cpp
// Search in a Sorted Infinite Array (Medium)

using namespace std;

#include <iostream>
#include <limits>
#include <vector>

class ArrayReader {
public:
  vector<int> arr;

  ArrayReader(const vector<int> &arr) { this->arr = arr; }

  virtual int get(int index) {
    if (index >= arr.size()) {
      return numeric_limits<int>::max();
    }
    return arr[index];
  }
};

class Solution {
public:
  static int searchInfiniteSortedArray(ArrayReader reader, int key) {
    int start = 0;
    int end = 1;

    while (reader.get(end) < key) {
        int newStart = end + 1;
        end += (end - start + 1) * 2; // increase the bounds size
        start = newStart;
    }

    return binarySearch(reader, key, start, end);
  }

private:
    static int binarySearch(ArrayReader reader, int key, int start, int end) {
        while (start <= end) {
            int mid = start + (end - start) / 2;
            if (key < reader.get(mid))
                end = mid - 1;
            else if (key > reader.get(mid))
                start = mid + 1;
            else 
                return mid;
        }

        return -1;
    }
};
